### PR TITLE
Add dr source code

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this PR, we refactor the data reshuffler (from system verilog to chisel) and enlarge the data reshuffler with max pool capability.

### CODE REVIEW

**Comment by rgantonio** [comment]
@xiaoling-yi  What is DR? data reshuffler? what is this addition? haha

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  // the receiver isn't ready, needs to send several cycles
+  val keep_output = RegInit(0.B)
+
+  when(io.data.input_i.valid) {
+    when(io.ctrl.init_i) {
+      // max pooling
+      resReg := Mux(io.data.input_i.bits > -128.S, io.data.input_i.bits, -128.S)
+    }.otherwise {
+      // max pooling
+      resReg := Mux(io.data.input_i.bits > resReg, io.data.input_i.bits, resReg)
```
Comment (rgantonio) [inline hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala:45]: Does this mean that  max pool returns the answer after N cycles? Where N is the number of elements to max pool?

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  val io = IO(new MAXPoolPEIO(params))
+
+  val resReg = RegInit(0.S(params.inputWidth.W))
+  // the receiver isn't ready, needs to send several cycles
+  val keep_output = RegInit(0.B)
+
+  when(io.data.input_i.valid) {
+    when(io.ctrl.init_i) {
+      // max pooling
+      resReg := Mux(io.data.input_i.bits > -128.S, io.data.input_i.bits, -128.S)
```
Comment (rgantonio) [inline hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala:42]: Why is this here? Like what is the `init_i` for? 

So when `init_i` it returns either the input or the lowest value?

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala
Comment (rgantonio) [inline hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala:1]: On 2nd thought, why is this here? hahahaha I think you meant this to be in the other PR yes?

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Parameter.scala [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  // input data type
+  def inputWidth = 8
+  // output data type
+  def outputWidth = 8
+
+  // Pooling parallelism
+  def poolLaneLen = 64
+
+  // we use 3 CSR for Reshuffler configuration,
```
Comment (rgantonio) [inline hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Parameter.scala:30]: 3 or 1? haha

Comment (rgantonio) [reply hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Parameter.scala:30]: oh wait 1 RW for the transpose or not

then 2 for perf counter and 1 busy. Is it still worth it to have a perf counter and busy if you do this in one cycle?

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+  }
+  io.performance_counter := performance_counter
+
+  config_valid := io.ctrl.fire
+
+  // when config valid, store the configuration for later computation
+  when(config_valid) {
+    ctrl_csr.opcode_i := io.ctrl.bits(0)(1, 0)
+    ctrl_csr.reduceLen_i := io.ctrl.bits(0)(6, 2)
+    ctrl_csr.tloop2Len_i := io.ctrl.bits(0)(31, 7)
```
Comment (rgantonio) [inline hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala:123]: Wait there are more features??? haha what are reduce len? what is also tloop? XD

Maybe give minimal documentation to PR on what is expected of this accelerator?

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala
Comment (rgantonio) [inline hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala:1]: So wait, the reshuffler has a max pooling combined with it too?

**Review by rgantonio — COMMENTED** [review]
I approve after discussion hahaha

**Comment by rgantonio** [comment]
Aha I saw the PR description after review. Wait!

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Parameter.scala (hunk 3 above)
Comment (xiaoling-yi) [reply hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Parameter.scala:30]: Yes!

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala
Comment (xiaoling-yi) [reply hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala:1]: yes. The share the TCDM ports

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala
Comment (rgantonio) [reply hw/chisel_acc/src/main/scala/snax_acc/reshuffle/MaxPoolingPE.scala:1]: Nvm haha after offline discussions I understand now.

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    lane.map(_.io.data.input_i.bits := 0.S)
+    lane.map(_.io.data.input_i.valid := false.B)
+    lane.map(_.io.data.output_o.ready := false.B)
+    result.map(_ := 0.S)
+  }
+
+  // for generating the init signal for the first MAXPoolPE
+  pe_input_fire := lane.map(_.io.data.input_i.fire).reduce(_ && _)
+  when(
+    pe_input_fire && pe_input_fire_counter =/= (ctrl_csr.reduceLen_i - 1.U) && cstate === sBUSY
```
Comment (rgantonio) [inline hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala:168]: Is there some kind of POSIX ordering here? Or is Chisel smart enough to know which to evaluate first? Like would it be better if:

`pe_input_fire && (pe_input_fire_counter =/= (ctrl_csr.reduceLen_i - 1.U)) && (cstate === sBUSY)`

**Review by rgantonio — APPROVED** [review]
Nice! Just answer then I approve!

File: hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala (hunk 5 above)
Comment (xiaoling-yi) [reply hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala:168]: It is
